drivers/mcp2515/mcp2515.c

```diff
@@ -0,0 +1,412 @@
+        frame->can_id = ((uint32_t)inbuf[0] << 3) +
+                        (((uint32_t)inbuf[1] & 0xE0) >> 5);
+    }
+
+    frame->can_dlc = inbuf[4];
+    memcpy(frame->data, inbuf + 5, frame->can_dlc);
+    return mailbox;
+}
+
+int mcp2515_abort(candev_mcp2515_t *dev, int mailbox)
+{
+    return mcp2515_spi_bitmod(dev, TX_CTRL(mailbox), MCP2515_TXBCTRL_TXREQ, 0);
+}
+
+enum mcp2515_mode mcp2515_get_mode(candev_mcp2515_t *dev)
+{
+    enum mcp2515_mode result = MODE_UNKNOWN;
+    uint8_t mode;
+
+    int res = mcp2515_spi_read(dev, MCP2515_CANSTAT, &mode, 1);
+
+    if (res == 0) {
+        result = (enum mcp2515_mode)mode & MCP2515_CANSTAT_OPMOD_MASK;
+    }
+
+    DEBUG("mcp2515_get_mode: mode=%x\n", result);
+
+    return result;
+}
+
+enum mcp2515_mode mcp2515_set_mode(candev_mcp2515_t *dev,
+                                   enum mcp2515_mode mode)
+{
+    DEBUG("mcp2515_set_mode: mode=%x\n", mode);
+
+    if (mode == MODE_UNKNOWN) {
+        return -1;
+    }
+
+    enum mcp2515_mode cur_mode = mcp2515_get_mode(dev);
+
+    if (cur_mode == mode) {
+        return mode;
+    }
+    if (cur_mode == MODE_SLEEP) {
+        mcp2515_wake_up(dev);
+    }
+
+    mcp2515_spi_bitmod(dev, MCP2515_CANCTRL, MCP2515_CANCTRL_REQOP_MASK,
+                       (uint8_t)mode);
+    int cnt = 0;
+
+    while (cur_mode != mode && cnt++ < 2) {
+        cur_mode = mcp2515_get_mode(dev);
+    }
+    return cur_mode;
+}
+
+void mcp2515_wake_up(candev_mcp2515_t *dev)
+{
+    dev->wakeup_src = MCP2515_WKUP_SRC_INTERNAL;
+    mcp2515_spi_bitmod(dev, MCP2515_CANINTF, MCP2515_CANINTF_WAKIF,
+                       MCP2515_CANINTF_WAKIF);
+    xtimer_usleep(osc_startup(dev));
+
+    uint8_t flag = mcp2515_get_irq(dev);
+
+    if (flag & INT_WAKEUP) {
+        DEBUG("wakeup, irq raised\n");
+        dev->wakeup_src = 0;
+        mcp2515_clear_irq(dev, INT_WAKEUP);
+    }
+}
+
+static int mcp2515_enable_irq(candev_mcp2515_t *dev, uint8_t irq)
+{
+    return mcp2515_spi_bitmod(dev, MCP2515_CANINTE, irq, irq);
+}
+
+int mcp2515_configure_bittiming(candev_mcp2515_t *dev)
+{
+    struct can_bittiming *tim = &(dev->candev.bittiming);
+    int res = 0;
+    uint8_t c;
+    enum mcp2515_mode mode;
+
+    DEBUG("mcp2515_configure_bittiming: brp=%" PRIu32 ", prop_seg=%" PRIu32
+          ", phase_seg1=%" PRIu32 ", phase_seg2=%" PRIu32 "\n", tim->brp,
+          tim->prop_seg,
+          tim->phase_seg1, tim->phase_seg2);
+
+    mode = mcp2515_get_mode(dev);
+    if (mode != MODE_CONFIG) {
+        if (mcp2515_set_mode(dev, MODE_CONFIG) != MODE_CONFIG) {
+            return -1;
+        }
+    }
+
+    c = ((tim->brp - 1) & 0x3F) | ((tim->sjw - 1) << 6);
+    mcp2515_spi_write(dev, MCP2515_CNF1, &c, 1);
+
+
+    mcp2515_spi_bitmod(dev, MCP2515_CNF2, MCP2515_CNF2_PRSEG_MASK |
+                       MCP2515_CNF2_PHSEG_MASK | MCP2515_CNF2_BTLMODE,
+                       MCP2515_CNF2_BTLMODE | (tim->prop_seg - 1) |
+                       ((tim->phase_seg1 - 1) << 3));
```

Comment:
Magic number?

Reply:
This is the offset for the PHSEG1[2:0] register in CNF2. 
In this case, I'm a bit hesitant to replace the magic number with a define, because it would be a unique case. 
I'm afraid it would cause more confusion than it would solve.

Reply:
OK, to bad there are no CMSIS stile headers